### ISSUE TITLE
A networkchaos caveat

### ISSUE THREAD

**Issue by hoang-tranviet** [chaos-mesh/chaos-mesh#1485]
Hello all,

It is good to see much progress of this project.

However, I have a concern about the loss emulation chaos.
As per [netem documentation](https://wiki.linuxfoundation.org/networking/netem#caveats), a caveat is: 

> When loss is used locally (not on a bridge or router), the loss is reported to the upper level protocols. This may cause TCP to resend and behave as if there was no loss. When testing protocol reponse to loss it is best to use a netem on a bridge or router 

I ran a test with different loss rates (with latency of ~300ms) and measured the file transfer time.

- For setting of 1% loss, `ss -it` shown no loss observed by TCP and there is no different in performance. 
- For setting of higher loss rate, the transfer is slower and there are retransmissions but still less than the loss level we set. 

For reference, https://progmp.net/mininetPitfalls.html discussed a pitfall with latency emulation due to _TCP small queues (TSQ)_ operation.

A solution I think of is to create a new local network device and set it as default device. The original eth0 will still be used for setting qdiscs, but now act as a proxy between the new device and external world. Not sure how difficult it is.

Thank you in advance.

**Comment by cwen0** [comment]
> Hello all,
> 
> It is good to see much progress of this project.
> 
> However, I have a concern about the loss emulation chaos.
> As per [netem documentation](https://wiki.linuxfoundation.org/networking/netem#caveats), a caveat is:
> 
> > When loss is used locally (not on a bridge or router), the loss is reported to the upper level protocols. This may cause TCP to resend and behave as if there was no loss. When testing protocol reponse to loss it is best to use a netem on a bridge or router
> 
> I ran a test with different loss rates (with latency of ~300ms) and measured the file transfer time.
> 
> * For setting of 1% loss, `ss -it` shown no loss observed by TCP and there is no different in performance.
> * For setting of higher loss rate, the transfer is slower and there are retransmissions but still less than the loss level we set.
> 
> For reference, https://progmp.net/mininetPitfalls.html discussed a pitfall with latency emulation due to _TCP small queues (TSQ)_ operation.
> 
> A solution I think of is to create a new local network device and set it as default device. The original eth0 will still be used for setting qdiscs, but now act as a proxy between the new device and external world. Not sure how difficult it is.
> 
> Thank you in advance.

Good catch! We will consider this issue carefully！

**Comment by cwen0** [comment]
@hoang-tranviet This loss emulation chaos is on TCP level. We are developing HTTPChaos and GRPCChao, and we will use a proxy to implement network loss on HTTP or Grpc level.

**Comment by hoang-tranviet** [comment]
I have set up a local network device in between (the bonding device specifically) but there was no difference in performance.

Then I tried changing `correlation` to 0% and the loss chaos actually worked!
It seems that netem loss does not work well with `correlation: 25%` (default value), at least on my kernel (5.4.0).

**Comment by YangKeao** [comment]
> @hoang-tranviet This loss emulation chaos is on TCP level. We are developing HTTPChaos and GRPCChao, and we will use a proxy to implement network loss on HTTP or Grpc level.

NO :imp: ! The loss emulation chaos works on layer 2, and that's why it has little effect for TCP (because the design of TCP has considered that the layer 2 is unstable and could loss packet).

**Comment by hoang-tranviet** [comment]
Yes, TCP will handle packet loss but with performance penalty when using cubic. But the kernel statistics (`tc -s qdisc`) show no packet drop when correlation = 25% and loss=1%.
